src/TraceEvent/EventPipe/EventPipeEventSource.cs

```diff
@@ -498,24 +535,25 @@ private void ReadAndCacheMetadataV6OrGreater(ref SpanReader reader)
         private void CacheMetadata(EventPipeMetadata metadata)
         {
             DynamicTraceEventData eventTemplate = CreateTemplate(metadata);
-            _eventMetadataDictionary.Add(metadata.MetaDataId, metadata);
-            _metadataTemplates[eventTemplate] = eventTemplate;
+            try
+            {
+                _eventMetadataDictionary.Add(metadata.MetaDataId, metadata);
+                _metadataTemplates[eventTemplate] = eventTemplate;
+            }
+            catch (ArgumentException)
+            {
+                EventPipeMetadata old = _eventMetadataDictionary[metadata.MetaDataId];
+                throw new FormatException($"Metadata ID {metadata.MetaDataId} already in use. Previously this id defined event {old.ProviderName}\\{old.EventName}. Now redefined for event {metadata.ProviderName}\\{metadata.EventName}.");
+            }
         }
 
         internal bool TryGetMetadata(int metadataId, out EventPipeMetadata metadata) =>  _eventMetadataDictionary.TryGetValue(metadataId, out metadata);
         internal bool TryGetThread(long threadIndex, out EventPipeThread thread) => _threadCache.TryGetValue(threadIndex, out thread);
 
-        private TraceEventNativeMethods.EVENT_RECORD* ConvertEventHeaderToRecord(ref EventPipeEventHeader eventData)
+        internal void FlushMetadataCache()
         {
-            if (_eventMetadataDictionary.TryGetValue(eventData.MetaDataId, out var metaData))
-            {
-                return metaData.GetEventRecordForEventData(eventData);
-            }
-            else
-            {
-                Debug.Assert(false, "Warning can't find metaData for ID " + eventData.MetaDataId);
-                return null;
-            }
+            _eventMetadataDictionary.Clear();
+            _metadataTemplates.Clear();
```

Comment:
Once the `_metadataTemplates` are cleared, would the `m_state.m_templates` dictionary need to be updated or completely flushed as well? I'm still getting acquainted with this codebase, but I was thinking that known metadata templates would get added to the `m_state.m_templates` dictionary through the `RegisterUnhandledEvent` callback https://github.com/microsoft/perfview/blob/a8605fe4c132052063025cf96ed51042dba95959/src/TraceEvent/RegisteredTraceEventParser.cs#L1185-L1224, and once we flush, they're still in that dictionary.